### ISSUE TITLE
Some questions about timestamp

### ISSUE THREAD

**Issue by hcv1027** [cyberbotics/webots_ros#20]
I have several question about simulation time.
1. I can't understand the effect of below sample code get from webots's ros example:
```c++
  ros::ServiceClient time_step_client = nh.serviceClient<webots_ros::set_int>(
      controllerName + "/robot/time_step");
  webots_ros::set_int time_step_srv;
  time_step_srv.request.value = TIME_STEP; // TIME_STEP = 16 in my case
  while (ros::ok()) {
    ros::spinOnce();
    if (!time_step_client.call(time_step_srv) ||
        !time_step_srv.response.success) {
      ROS_ERROR("Failed to call service time_step for next step.");
    }
    // Do something here.
  }
``` 
I can't find any change as I add or remove the `if (...)` block.

2. The message's timestamp I get from ros topic looks like real time timestamp, not webots's simulation timestamp. For example, in below image, the webots's simulation time only passed about 3 seconds, but the timestamp of the message my controller received passed about 9.82 seconds. And I use my watch to confirm that this 9.82 seconds is equal to the passed time since webots is ready.
![timestamp](https://user-images.githubusercontent.com/12728923/84851149-17d58280-b08c-11ea-8132-c12068090f8c.png)

I also try to use `<param name="/use_sim_time" value="true"/>` in my launch file, and add the `if (...)` block in my while loop, but the webot's simulation just hold there, the timestamp will not change. How can I receive the timestamp which is equivalent to the timestamp in webots' simulation?

**Comment by DavidMansolino** [comment]
By default the simulation is not synchronized with your ros node, whcih means simulation will run independently from your ros node, when you call for the first time the `/robot/time_step` service, it will start the synchronization, and Webots will then wait that you call again this service to advance one simulation step.

About the timestamp, you should add the `--clock` and `--use-sim-time` to the ros controller in Webots: https://cyberbotics.com/doc/guide/using-ros#standard-ros-controller

**Comment by hcv1027** [comment]
Hi @DavidMansolino ,
Can you give me some example of how and where to use these ros controller parameters?
Here is my current launch file:
```
<?xml version="1.0"?>
<launch>
  <arg name="use_joystick" default="true" />
  <param name="/use_sim_time" value="true"/>

  <!-- start Webots -->
  <arg name="no-gui" default="false," doc="Start Webots with minimal GUI"/>
  <include file="$(find webots_ros)/launch/webots.launch">
    <arg name="mode" value="pause"/>
    <arg name="no-gui" value="$(arg no-gui)"/>
    <arg name="world" value="$(find webots_office)/worlds/simple_world.wbt"/>
    <arg name="clock" value="/clock"/>
    <arg name="use-sim-time" value="true"/>
  </include>

  <!-- Publish tf transform -->
  <node pkg="tf" type="static_transform_publisher" name="lidar_broadcaster"
    args="0 0 0.1185 0 0 -3.14159 base_link eaibot/lidar 100" />
  <node pkg="tf" type="static_transform_publisher" name="footprint_broadcaster"
    args="0 0 -0.0735 0 0 0 1 base_link base_footprint 100" />

  <arg name="auto-close" default="false" doc="Startup mode"/>
  <node name="webots_office" pkg="webots_office" type="webots_office_node"
    required="$(arg auto-close)" output="screen">
  </node>

  <include file="$(find nuwa_robot)/launch/hardware/joystick.launch"
    if="$(arg use_joystick)" />
</launch>

```

I try to use them in my launch file
```
<!-- start Webots -->
  <arg name="no-gui" default="false," doc="Start Webots with minimal GUI"/>
  <include file="$(find webots_ros)/launch/webots.launch">
    <arg name="mode" value="pause"/>
    <arg name="no-gui" value="$(arg no-gui)"/>
    <arg name="world" value="$(find webots_office)/worlds/simple_world.wbt"/>
    <arg name="clock" value="/clock"/>
    <arg name="use-sim-time" value="true"/>
  </include>
```
But webots will show error message like this: 
```
unused args [use-sim-time, clock] for include of [/opt/ros/kinetic/share/webots_ros/launch/webots.launch]
```


And I use the command `rosservice list`, there is also no service for me to set these argments.

**Comment by DavidMansolino** [comment]
These arguments should be set directly in the `controllerArgs` fields of your robot in the Webots scene-tree, see for example:
https://github.com/cyberbotics/webots/blob/510730f9d80b666eb3b04b1988e36e470f1f9977/projects/languages/ros/worlds/complete_test.wbt#L270

**Comment by hcv1027** [comment]
Hi @DavidMansolino , after I set these arguments like below image, the service call just blocks there. It will not return.
![robot_setting](https://user-images.githubusercontent.com/12728923/84991442-19c93f80-b179-11ea-8156-5ea4220a7cde.png)

In `my_ros_node.cpp`, the log will only show `time_step_client++`, then blocks at the call `time_step_client.call(time_step_srv);`. All kind of service call has the same problem.
```c++
  ros::ServiceClient time_step_client = nh.serviceClient<webots_ros::set_int>(
      controllerName + "/robot/time_step");
  webots_ros::set_int time_step_srv;
  time_step_srv.request.value = TIME_STEP;
  ROS_INFO("time_step_client++");
  time_step_client.call(time_step_srv);
  ROS_INFO("time_step_client--");
```

**Comment by DavidMansolino** [comment]
Is the simulation running?
Have you tried removing some arguments (such as `--use-sim-time`for example)?

**Comment by hcv1027** [comment]
Hi @DavidMansolino , after testing, if I remove the `<param name="/use_sim_time" value="true"/>` in my launch file, the webots simulation will run. But the time still does not synchronize with ROS. As long as I set ROS parameter `/use_sim_time` to true, all webots service call will be blocked, and the simulations will not run (time keeps 0:00:00 like below image)
![time_stop](https://user-images.githubusercontent.com/12728923/85087998-94459e00-b211-11ea-8858-19f76dd266bb.png)
.

**Comment by DavidMansolino** [comment]
And from the controller args in Webots, have you tried to remove the `--use-sim-time` arguments?

**Comment by hcv1027** [comment]
For clarity, I use the table to show my testing:
![test](https://user-images.githubusercontent.com/12728923/85109204-89a3fc80-b243-11ea-9184-58745fb50313.png)
The keypoint is whether I set `/use_sim_time` to `true` or `false` in ROS launch file. If I set true, the webots will be blocked, no matter whether I add or remove `--use-sim-time arguments` in webots. And the time doesn't synchronize between webots and ROS if I only add `--use-sim-time arguments` in webots and set `/use_sim_time` to `false` in ROS launch file.
